0000-improved-flow-types-for-babel-types.md

```diff
@@ -0,0 +1,174 @@
+* Change the `%checks` on the `is*` function declarations from `node instanceof XNode` to `node != null && node.type === 'X'`
+
+Example Usage:
+
+```js
+import * as t from '@babel/types';
+import type { StringLiteral, NumericLiteral } from '@babel/types';
+import { isStringLiteral } from '@babel/types`;
+
+function asStringLiteral(ast: ?StringLiteral | NumericLiteral): StringLiteral | null {
+  if (ast == null) {
+      return t.stringLiteral("null");
+  }
+
+  if (isStringLiteral(ast)) { // or ast.type === "StringLiteral"
+    return ast;
+  }
+
+  return t.stringLiteral(String(ast.value));
+}
+```
+
+This changes can be implemented by changing the type definition generation script in `babel/types` for flow only.
+
+## Regression
+* The node types can no longer be referenced in `%check` declarations. E.g. `%checks (node instanceof BabelNodeIdentifier)` is no longer valid because `BabelNodeIdentifier` is now a type and not a class. The refinement must be rewritten to `%checks (node != null && node.type ==='Identifier')`
+* Referencing node types in a value position is a compile error. For example `import {BabelNodeIdentifier} from '@babel/types';` is no longer valid because `@babel/types` does not export such value. It must be changed to `import type {BabelNodeIdentifier} from '@babel/types'`.
+
+
+# Drawbacks
+
+* Braking Change:
+    * References to the node types as values must be rewritten to type references
+    * `%checks` declaration in consumer code must be rewritten to refine on `type`.
+* Unions can potentially be expensive during type checking. Flow does some optimisations and I tried out the my definition file and haven't noticed a regression compared to the existing type definition.
+* Union type error messages can be hard to understand because flow lists each possible option.
+
+# Alternatives
+
+## Keep as is
+
+The drawbacks are:
+
+* Having to import the `is*` methods for type refinement to work and there's no alternative way to refine nodes.
+* Defining the node types as `class`es is misguiding for users that the node types are modelled as values (`class`es) when they're not
+* There's no way to refine a `path` because Flow doesn't support the refinement of the `path.isX` methods and refining on `path.node.type` isn't sufficient because `node` isn't defined as a union.
+
+## Inexact vs Exact Types
+
+An alternative is to define the node types either as inexact types or interfaces. The difference is this would allow consumers to set additional props on the nodes.
+
+I believe that exact types are the right choice here since additional properties are e.g. not supported by `babel/generate` and `@babel/traverse` offers `NodePath.setData/getData` to store custom data.
+
+# Adoption strategy
+
+## Create a transform script
+
+Create a babel plugin that rewrites the `%checks` refinements and changes the imports of node types from values to types.
```

Comment:
I don't think that this will be need in practice. Personally I expect that the number of people manually writing `%checks` for Babel types in custom functions is close to zero, and I would consider this PR strictly as a bugfix (i.e. mark nodes not as classes).

Code like this is invalid in flow:
```js
function check(x: A | B): %checks (x instanceof A) {
    return x.type === "a";
}
```
[repl](https://flow.org/try/#0JYOwLgpgTgZghgYwgAgILIN4Chm+WATwAcIAuZAIjgqwF8stRJZEUAhTHPQk8igIwoBuOgxgBXEAjDAA9iGQIAFhAQBrABQAPcugA+yNgEpyAUmWq1AZ2TaAdDxQBeF5WpHOeZFAhhxUBS0HYmdXKmFRIA)